File: packages/backend/src/misc/check-hit-antenna.ts
import type { Antenna } from "@/models/entities/antenna.js";
import type { Note } from "@/models/entities/note.js";
import type { User } from "@/models/entities/user.js";
import type { UserProfile } from "@/models/entities/user-profile.js";
import { Blockings, Followings, UserProfiles } from "@/models/index.js";
import { checkWordMute, getFullApAccount, stringToAcct } from "backend-rs";
import type { Packed } from "@/misc/schema.js";
import { Cache } from "@/misc/cache.js";

const blockingCache = new Cache<User["id"][]>("blocking", 60 * 5);
const hardMutesCache = new Cache<{
	userId: UserProfile["userId"];
	mutedWords: UserProfile["mutedWords"];
	mutedPatterns: UserProfile["mutedPatterns"];
}>("hardMutes", 60 * 5);
const followingCache = new Cache<User["id"][]>("following", 60 * 5);

export async function checkHitAntenna(
	antenna: Antenna,
	note: Note | Packed<"Note">,
	noteUser: { id: User["id"]; username: string; host: string | null },
): Promise<boolean> {
	if (note.visibility === "specified") return false;
	if (antenna.withFile) {
		if (note.fileIds && note.fileIds.length === 0) return false;
	}
	if (!antenna.withReplies && note.replyId != null) return false;

	if (antenna.src === "users") {
		const accts = antenna.users.map((x) => {
			const { username, host } = stringToAcct(x);
			return getFullApAccount(username, host).toLowerCase();
		});
		if (
			!accts.includes(
				getFullApAccount(noteUser.username, noteUser.host).toLowerCase(),
			)
		)
			return false;
	} else if (antenna.src === "instances") {
		const instances = antenna.instances
			.filter((x) => x !== "")
			.map((host) => {
				return host.toLowerCase();
			});
		if (!instances.includes(noteUser.host?.toLowerCase() ?? "")) return false;
	}

	const keywords = antenna.keywords
		// Clean up
		.map((xs) => xs.filter((x) => x !== ""))
		.filter((xs) => xs.length > 0);

	let text = `${note.text ?? ""} ${note.cw ?? ""}`;
	if (note.files != null)
		text += ` ${note.files.map((f) => f.comment ?? "").join(" ")}`;
	text = text.trim();

	if (keywords.length > 0) {
		if (note.text == null) return false;

		const matched = keywords.some((and) =>
			and.every((keyword) =>
				antenna.caseSensitive
					? text.includes(keyword)
					: text.toLowerCase().includes(keyword.toLowerCase()),
			),
		);

		if (!matched) return false;
	}

	const excludeKeywords = antenna.excludeKeywords
		// Clean up
		.map((xs) => xs.filter((x) => x !== ""))
		.filter((xs) => xs.length > 0);

	if (excludeKeywords.length > 0) {
		if (note.text == null) return false;

		const matched = excludeKeywords.some((and) =>
			and.every((keyword) =>
				antenna.caseSensitive
					? note.text?.includes(keyword)
					: note.text?.toLowerCase().includes(keyword.toLowerCase()),
			),
		);

		if (matched) return false;
	}

	// アンテナ作成者がノート作成者にブロックされていたらスキップ
	const blockings = await blockingCache.fetch(noteUser.id, () =>
		Blockings.findBy({ blockerId: noteUser.id }).then((res) =>
			res.map((x) => x.blockeeId),
		),
	);
	if (blockings.includes(antenna.userId)) return false;

	if (note.visibility === "followers" || note.visibility === "home") {
		const following = await followingCache.fetch(antenna.userId, () =>
			Followings.find({
				where: { followerId: antenna.userId },
				select: ["followeeId"],
			}).then((relations) => relations.map((relation) => relation.followeeId)),
		);
		if (!following.includes(note.userId)) return false;
	}

	const mutes = await hardMutesCache.fetch(antenna.userId, () =>
		UserProfiles.findOneByOrFail({
			userId: antenna.userId,
		}).then((profile) => {
			return {
				userId: antenna.userId,
				mutedWords: profile.mutedWords,
				mutedPatterns: profile.mutedPatterns,
			};
		}),
	);
	if (
		mutes.mutedWords != null &&
		mutes.mutedPatterns != null &&
		antenna.userId !== note.userId &&
		(await checkWordMute(note, mutes.mutedWords, mutes.mutedPatterns))
	)
		return false;

	// TODO: eval expression

	return true;
}

File: packages/backend-rs/src/misc/check_hit_antenna.rs
use crate::config::CONFIG;
use crate::database::{cache, db_conn};
use crate::federation::acct::Acct;
use crate::misc::get_note_all_texts::{all_texts, NoteLike};
use crate::model::entity::{antenna, blocking, following, note, sea_orm_active_enums::*};
use sea_orm::{ColumnTrait, DbErr, EntityTrait, QueryFilter, QuerySelect};

#[derive(thiserror::Error, Debug)]
pub enum AntennaCheckError {
    #[error("Database error: {0}")]
    DbErr(#[from] DbErr),
    #[error("Cache error: {0}")]
    CacheErr(#[from] cache::Error),
    #[error("User profile not found: {0}")]
    UserProfileNotFoundErr(String),
}

fn match_all(space_separated_words: &str, text: &str, case_sensitive: bool) -> bool {
    if case_sensitive {
        space_separated_words
            .split_whitespace()
            .all(|word| text.contains(word))
    } else {
        space_separated_words
            .to_lowercase()
            .split_whitespace()
            .all(|word| text.to_lowercase().contains(word))
    }
}

pub async fn check_hit_antenna(
    antenna: &antenna::Model,
    note: note::Model,
    note_author: &Acct,
) -> Result<bool, AntennaCheckError> {
    if note.visibility == NoteVisibilityEnum::Specified {
        return Ok(false);
    }

    if antenna.with_file && note.file_ids.is_empty() {
        return Ok(false);
    }

    if !antenna.with_replies && note.reply_id.is_some() {
        return Ok(false);
    }

    if antenna.src == AntennaSrcEnum::Users {
        let is_from_one_of_specified_authors = antenna
            .users
            .iter()
            .map(|s| s.parse::<Acct>().unwrap())
            .any(|acct| acct.username == note_author.username && acct.host == note_author.host);

        if !is_from_one_of_specified_authors {
            return Ok(false);
        }
    } else if antenna.src == AntennaSrcEnum::Instances {
        let is_from_one_of_specified_servers = antenna.instances.iter().any(|host| {
            host.to_ascii_lowercase()
                == note_author
                    .host
                    .clone()
                    .unwrap_or(CONFIG.host.clone())
                    .to_ascii_lowercase()
        });

        if !is_from_one_of_specified_servers {
            return Ok(false);
        }
    }

    // "Home", "Group", "List" sources are currently disabled

    let note_texts = all_texts(NoteLike {
        file_ids: note.file_ids,
        user_id: note.user_id.clone(),
        text: note.text,
        cw: note.cw,
        renote_id: note.renote_id,
        reply_id: note.reply_id,
    })
    .await?;

    let has_keyword = antenna.keywords.iter().any(|words| {
        note_texts
            .iter()
            .any(|text| match_all(words, text, antenna.case_sensitive))
    });

    if !has_keyword {
        return Ok(false);
    }

    let has_excluded_word = antenna.exclude_keywords.iter().any(|words| {
        note_texts
            .iter()
            .any(|text| match_all(words, text, antenna.case_sensitive))
    });

    if has_excluded_word {
        return Ok(false);
    }

    let db = db_conn().await?;

    let blocked_user_ids: Vec<String> = cache::get_one(cache::Category::Block, &note.user_id)?
        .unwrap_or({
            // cache miss
            let blocks = blocking::Entity::find()
                .select_only()
                .column(blocking::Column::BlockeeId)
                .filter(blocking::Column::BlockerId.eq(&note.user_id))
                .into_tuple::<String>()
                .all(db)
                .await?;
            cache::set_one(cache::Category::Block, &note.user_id, &blocks, 10 * 60)?;
            blocks
        });

    // if the antenna owner is blocked by the note author, return false
    if blocked_user_ids.contains(&antenna.user_id) {
        return Ok(false);
    }

    if [NoteVisibilityEnum::Home, NoteVisibilityEnum::Followers].contains(&note.visibility) {
        let following_user_ids: Vec<String> =
            cache::get_one(cache::Category::Follow, &antenna.user_id)?.unwrap_or({
                // cache miss
                let following = following::Entity::find()
                    .select_only()
                    .column(following::Column::FolloweeId)
                    .filter(following::Column::FollowerId.eq(&antenna.user_id))
                    .into_tuple::<String>()
                    .all(db)
                    .await?;
                cache::set_one(
                    cache::Category::Follow,
                    &antenna.user_id,
                    &following,
                    10 * 60,
                )?;
                following
            });

        // if the antenna owner is not following the note author, return false
        if !following_user_ids.contains(&note.user_id) {
            return Ok(false);
        }
    }

    Ok(true)
}

#[cfg(test)]
mod unit_test {
    use super::match_all;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_match_all() {
        assert_eq!(match_all("Apple", "apple and banana", false), true);
        assert_eq!(match_all("Apple", "apple and banana", true), false);
        assert_eq!(match_all("Apple Banana", "apple and banana", false), true);
        assert_eq!(match_all("Apple Banana", "apple and cinnamon", true), false);
        assert_eq!(
            match_all("Apple Banana", "apple and cinnamon", false),
            false
        );
    }
}
